pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+	return nil
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Start() error {
+	appMgr.controller = NewController(appMgr)
+	appMgr.controller.Run()
+	return nil
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Stop() {
+	// noop
+}
+
+// handle the updates from the scheduler and sync the status change
+func (appMgr *AppManager) HandleCallbackEvents() func(obj interface{}) {
+	return func(obj interface{}) {
+		if event, ok := obj.(events.ApplicationEvent); ok {
+			if events.AppStateChange == event.GetEvent() {
+				if shimEvent, ok := event.(shimcache.ApplicationStatusChangeEvent); ok {
+					appID := event.GetApplicationID()
+					log.Logger.Info("Status Change callback received",
+						zap.String("app id", appID),
+						zap.String("new status", shimEvent.State))
+					var app = appMgr.amProtocol.GetApplication(appID).(*shimcache.Application)
+					appCRD, err := appMgr.controller.lister.Applications(app.GetTags()[constants.AppTagNamespace]).Get(appID)
+					if err != nil {
+						log.Logger.Warn("Failed to query app CRD for status update",
+							zap.String("Application ID", appID),
+							zap.Error(err))
+						return
+					}
+					crdState := convertShimAppStateToAppCRDState(shimEvent.State)
+					if crdState != "Undefined" {
+						appMgr.updateAppCRDStatus(appCRD, crdState)
+					} else {
+						log.Logger.Error("Invalid status, skip saving it",
+							zap.String("App id", appID))
+					}
+				}
+			}
+		}
+	}
+}
+
+type Controller struct {
+	kubeClient       kubernetes.Interface
+	extensionsClient apiextensionsclient.Interface
+	appClient        applicationclient.Interface
+	informer         cache.SharedIndexInformer
+	lister           applister.ApplicationLister
+}
+
+func NewController(appMgr *AppManager) *Controller {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		log.Logger.Fatal(err.Error())
+	}
+
+	kubeClient := kubernetes.NewForConfigOrDie(config)
+	apiextensionsClient := apiextensionsclient.NewForConfigOrDie(config)
+	appClient := applicationclient.NewForConfigOrDie(config)
+
+	informerFactory := appinformers.NewSharedInformerFactory(appClient, time.Minute*1)
+	informer := informerFactory.Apache().V1alpha1().Applications()
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    appMgr.addApp,
+		UpdateFunc: appMgr.updateApp,
+		DeleteFunc: appMgr.deleteApp,
+	})
+	informerFactory.Start(wait.NeverStop)
+
+	utilruntime.Must(appv1.AddToScheme(appscheme.Scheme))
+	return &Controller{
+		kubeClient:       kubeClient,
+		extensionsClient: apiextensionsClient,
+		appClient:        appClient,
+		informer:         informer.Informer(),
+		lister:           informer.Lister(),
+	}
+}
+
+func (c *Controller) Run() {
+	//TODO: handle the case when the crd is not registered: register it programatically
+	defer utilruntime.HandleCrash()
```

Comment:
I am not quite sure what does this mean.
why do we need to handle the case the CRD is not registered? If the CRD is not registered, I think we should not even start the scheduler. And also, what does the `HandleCrash()` do here?

Reply:
Right now we have two places where we can enable the CRD: the scheduler conf and the helm chart. In my opinion we should reduce it to only one place, the scheduler conf. If we will kepp this two places, I think it will be a frequent user error that they will enable it in one place and forget to enable it in the helm chart for example, or will forget to register it manually. I think in this case it will be useful to register it from the code.

Reply:
OK, if this is the common pattern, then that is fine.

Reply:
It is not a common pattern, it is just easy to forget to enable it in both places. I created a different Jira for it: YUNIKORN-346